### ISSUE TITLE
not able to connect external domains when specified multiple service backends of external domains for closed mesh

### ISSUE THREAD

**Issue by suyog-fp** [aws/aws-app-mesh-controller-for-k8s#433]
**Describe the bug**
I have implemented default close mesh. I am trying to connect below 2 https sites 
https://github.com
https://stackoverflow.com
when I specify single backends for VirtualNode I am able to connect to any of above domain(by curl) but when I specify multiple backends like below for VirtualNode then both domains cant connect.
```
  backends:
    - virtualService:
        virtualServiceRef:
          name: github.com
    - virtualService:
        virtualServiceRef:
          name: stackoverflow.com
```
**Steps to reproduce**
The target virtual node yaml associated with app( from which I run curl)

```
apiVersion: appmesh.k8s.aws/v1beta2
kind: VirtualNode
metadata:
  name: deathstar-vnode
  namespace: my-apps-closed
spec:
  podSelector:
    matchLabels:
      app: deathstar
  listeners:
    - portMapping:
        port: 80
        protocol: http
  serviceDiscovery:
    dns:
      hostname: deathstar.my-apps-closed.svc.cluster.local
  backends:
    - virtualService:
        virtualServiceRef:
          name: github.com
    - virtualService:
        virtualServiceRef:
          name: stackoverflow.com
```

github-node-svc
```
---
apiVersion: appmesh.k8s.aws/v1beta2
kind: VirtualNode
metadata:
  name: github-node
  namespace: my-apps-closed
spec:
  listeners:
    - portMapping:
        port: 443
        protocol: tcp
  serviceDiscovery:
    dns:
      hostname: 'github.com'
---
apiVersion: appmesh.k8s.aws/v1beta2
kind: VirtualService
metadata:
  name: github.com
  namespace: my-apps-closed
spec:
  provider:
    virtualNode:
      virtualNodeRef:
        name: github-node
---
```

stack-node-svc
```
---
apiVersion: appmesh.k8s.aws/v1beta2
kind: VirtualNode
metadata:
  name: stack-node
  namespace: my-apps-closed
spec:
  listeners:
    - portMapping:
        port: 443
        protocol: tcp
  serviceDiscovery:
    dns:
      hostname: 'stackoverflow.com'

---
apiVersion: appmesh.k8s.aws/v1beta2
kind: VirtualService
metadata:
  name: stackoverflow.com
  namespace: my-apps-closed
spec:
  provider:
    virtualNode:
      virtualNodeRef:
        name: stack-node
---

```

**Expected outcome**
curl to both domain should work with 200 response code
```
curl -I https://github.com
curl -I https://stackoverflow.com
```

**Environment**

* App Mesh controller version - 1.2.0
* Envoy version - v1.15.1.0-prod
* Are you using any integrations? X-ray, Jaeger etc. If so versions? X-ray latest
* Kubernetes version v1.16.8-eks-e16311
* Using EKS (yes/no), if so version? v1.16.8-eks-e16311

**Additional Context**:
when both backends specified only lb_endpoints are created and filter_chains for those domains are not created in envoy config.

when i just specify single backend(just github.com) then filter_chains are created like below
```
       "filter_chains": [
        {
         "filter_chain_match": {
          "destination_port": 443
         },
         "filters": [
          {
           "name": "envoy.tcp_proxy",
           "config": {
            "stat_prefix": "egress",
            "cluster": "cds_egress_sandbox-mesh-aws-closed_github-node_my-apps-closed_tcp_443"
           }
          }
         ]
        },
```

**Comment by suyog-fp** [comment]
hi @cgchinmay @fawadkhaliq 
Any updates about this issue?

**Comment by cgchinmay** [comment]
Hi, haven't got a chance to look at it. Will take a look and get back to
you soon.

**Comment by cgchinmay** [comment]
Hi @suyog-fp Could you also upload your spec for creating the default close mesh ?

**Comment by suyog-fp** [comment]
hi @cgchinmay 
```

**Comment by suyog-fp** [comment]
`

**Comment by cgchinmay** [comment]
Hi @suyog-fp I have few more checks to do before I can give you a final update. Meanwhile, I would suggest you to check following steps:

- Check Controller and envoy logs and ensure that there are no errors related to fetching virtualService
- If you do see any errors then ensure that your worker nodes have the necessary[ IAM permissions](https://github.com/aws/eks-charts/blob/master/stable/appmesh-controller/README.md#setup-iam-permissions-manually-on-worker-nodes) applied 
- For your reference: here is an example walkthrough which demonstrates accessing an [external service](https://github.com/aws/aws-app-mesh-examples/tree/master/walkthroughs/howto-k8s-egress)

**Comment by rajal-amzn** [comment]
@suyog-fp In order to connect to external domains, defining a Virtual Service without a provider should be enough. Virtual Node is used to define the configuration of a service so that when an Envoy is connected as a proxy to the service, it is able to route traffic appropriately.

In your case, since you are trying to connect to an external service, just declaring that as `VirtualService` and referring them as backend should be enough.

So, you would modify as below:
```
apiVersion: appmesh.k8s.aws/v1beta2
kind: VirtualService
metadata:
  awsName: stackoverflow.com
  name: stackoverflow.com
  namespace: my-apps-closed
```

**Comment by cgchinmay** [comment]
Thanks @rajal-amzn  for helping out. @suyog-fp The main issue here is communicating over TLS and not multiple service backends. Your existing approach also works if it wasn't TLS. You will have to specify a [clientpolicy ](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appmesh-virtualnode-clientpolicytls.html)with TLS validation context in your virtual node to get 200 response code.

**Comment by suyog-fp** [comment]
> Hi @suyog-fp I have few more checks to do before I can give you a final update. Meanwhile, I would suggest you to check following steps:
> 
> * Check Controller and envoy logs and ensure that there are no errors related to fetching virtualService
> * If you do see any errors then ensure that your worker nodes have the necessary[ IAM permissions](https://github.com/aws/eks-charts/blob/master/stable/appmesh-controller/README.md#setup-iam-permissions-manually-on-worker-nodes) applied
> * For your reference: here is an example walkthrough which demonstrates accessing an [external service](https://github.com/aws/aws-app-mesh-examples/tree/master/walkthroughs/howto-k8s-egress)

hi @cgchinmay 
Thanks for checking but as I said before with single service backend the setup is working so it rules out IAM issues or other app mesh basic setup issue in my opinion. This issue is specifically with multiple **tcp** backends.Envoy is not able to correctly add multiple filter_chains when we specify more than 1 tcp backend.Please check the envoy configuration translation logic for it.

 * Check Controller and envoy logs and ensure that there are no errors related to fetching virtualService
-> Checked. could not find anything.

 * If you do see any errors then ensure that your worker nodes have the necessary[ IAM permissions](https://github.com/aws/eks-charts/blob/master/stable/appmesh-controller/README.md#setup-iam-permissions-manually-on-worker-nodes) applied
-> Already there

 * For your reference: here is an example walkthrough which demonstrates accessing an [external service](https://github.com/aws/aws-app-mesh-examples/tree/master/walkthroughs/howto-k8s-egress)
-> I referred this for my setup initially

**Comment by suyog-fp** [comment]
hi @cgchinmay @rajal-amzn 
specifying `awsName` fixed my other http backend issue :) 

for tcp multiple backends I tried clientpolicies setup but it wasn't working somehow that time so I removed it.
Here is thread related to that discussion/issue
https://awsappmesh.slack.com/archives/C0113RB44C9/p1611657084015000

can you let me know how I should specify it?

**Comment by rajal-amzn** [comment]
Thanks for the context. I now see why you have Virtual Node defined for TCP backends. As mentioned in the Slack channel, you do not need to add client Policy. I haven't tried external connections via TCP. Could you share the error message from the Envoy logs?

**Comment by cgchinmay** [comment]
> hi @cgchinmay @rajal-amzn
> specifying `awsName` fixed my other http backend issue but for tcp I still get same issue :)
> 
> for tcp multiple backends I tried clientpolicies setup but it wasn't working somehow that time so I removed it.
> Here is thread related to that discussion/issue
> https://awsappmesh.slack.com/archives/C0113RB44C9/p1611657084015000
> 
> can you let me know how I should specify it?

Hi @suyog-fp Thanks for sharing this discussion. You should have probably shared it when you filed the Issue. Anyway I went through the discussion and this reply https://awsappmesh.slack.com/archives/C0113RB44C9/p1611776717022500?thread_ts=1611657084.015000&cid=C0113RB44C9 looks pretty clear. I am not sure why it didn't work for you, could be because of some other reason. Let me give this a try and I will get back to you :)

**Comment by efe-selcuk** [comment]
We have a limitation on routing TCP backends which slipped my mind when I was discussing this in Slack with you @suyog-fp, sorry for the confusion. If multiple TCP backends are listening on the same port (e.g., 443 like in this case) we can't actually making a routing decision, because the port is the only info we can route with. That's why you (helpfully) noticed that the filter match no longer exists after adding the second backend.

So, using http backends is the solution here. And adding the TLS client policy to the client (deathstar) gives the proxy the additional context to validate the peer's TLS certificate. You will need to have the "application" (curl) use the http scheme and let Envoy handle encrypting the traffic.

This surfaced a separate issue which Chinmay mentioned above, in that if the endpoint is sending a 301 redirect to upgrade to https, that won't be handled correctly and is a bug. The proxy is not handling the redirect, as we don't [enable cross-scheme redirects](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http/http_connection_management#internal-redirects). And if you follow the redirect at the application, you're back to the issue of trying to originate TLS at the application.

So, use http backends and a client policy. If the endpoint you're trying to reach sends a redirect, that'll be a problem.

**Comment by cgchinmay** [comment]
@suyog-fp Let us know if you have any questions if not, I will close this issue. What you are asking for should be a feature request. You can link this issue and create a new one over here: https://github.com/aws/aws-app-mesh-roadmap/issues

**Comment by suyog-fp** [comment]
hi @cgchinmay @efe-selcuk @rajal-amzn 
I tried the suggestions by you. Here is summary
- Previously(many days back) when I tried clientPolicy for http,it failed because there was no `awsName` specified as domain in virtualService. It was using default one with namespace appended to it which caused connection timeout failure in envoy.
- Right now I tried below `http` backends (port 80 for www.recipepuppy.com and port 443 for github.com,stackoverflow.com)
```
    - virtualService:
        virtualServiceRef:
          name: github.com
    - virtualService:
        virtualServiceRef:
          name: stackoverflow.com
    - virtualService:
        virtualServiceRef:
          name: www.recipepuppy.com
```
in this I am getting 503 response for http://www.recipepuppy.com. github.com returns `200` and stackoverflow.com returns `301` so this is one more issue when we keep `enforce: true` in clientPolicy.
If I try  `enforce: false` then http://www.recipepuppy.com returns 200 and  github.com,stackoverflow.com returns `503`

in conclusion there are 2 bugs
- 301 redirect to upgrade to https
- http 80 port backends not supported when specified clientPolicy with `enforce: true`

These issues are more related to envoy so should I report above bugs in envoy project instead of here?

**Comment by cgchinmay** [comment]
enforce: false works means its not using https. In that case you wouldn't need client policy. 

You can also define client  policy under individual virtual services instead of backend defaults. It will work for recipepuppy.com and github.com but stackoverflow will still be an issue.

**Comment by suyog-fp** [comment]
@cgchinmay yeah! specifying clientPolicy for just nonhttps backend works :) so there is only 1 issue of http redirect remaining now which I created here https://github.com/aws/aws-app-mesh-roadmap/issues/317
